### PR TITLE
Accelerate ROIPooling layer

### PR DESCRIPTION
**Pull request by arcadiaphy**
## Description ##
As title. The major changes in implementation:

1. Use index_t to store max_idx,  avoiding possible value change when casting int to float.
2. Remove assert in code, fix a bug in processing invalid box
3. Store global index of input data blob in max_idx to simplify grad accumulation in backward pass.

### Benchmarking script ###
```
import mxnet as mx
import numpy as np
np.random.seed(0)
import time

batch_size = 4
channel = 16
height = 256
width = 256
n_rois = 500
pooled_size = (5, 5)

def test(ctx, tries=5):
    s = 0.
    for idx in xrange(tries + 1):
        x = np.random.random((batch_size, channel, height, width))
        y = np.zeros((n_rois, 5))
        y[:, 0] += np.floor(np.random.random((n_rois)) * batch_size)
        y[:, 1] += np.random.random((n_rois)) * width
        y[:, 2] += np.random.random((n_rois)) * height
        y[:, 3] += np.minimum(y[:, 1] + np.random.random((n_rois)) * width, width)
        y[:, 4] += np.minimum(y[:, 2] + np.random.random((n_rois)) * height, height)

        with ctx:
            x = mx.nd.array(x)
            y = mx.nd.array(y)
            mx.nd.waitall()
            start = time.time()
            with mx.autograd.record():
                x.attach_grad()
                r = mx.nd.ROIPooling(data=x, rois=y,
                                     spatial_scale=1,
                                     pooled_size=pooled_size)
                r.backward(mx.nd.ones_like(r))
            mx.nd.waitall()
            if idx > 0:
                s += time.time() - start
    print 'time: {} s on {}'.format(s / tries, ctx)

if __name__ == '__main__':
    test(mx.cpu())
    test(mx.gpu())
```

### Result ###
Before:
```
time: 13.2959038258 s on cpu(0)
time: 0.0346892356873 s on gpu(0)
```

After:
```
time: 0.6734582901 s on cpu(0)
time: 0.0038733959198 s on gpu(0)
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by karan6181** [comment]
@arcadiaphy Thanks for your contribution!
@sxjscience , @KellenSunderland Could you please review this PR? Thanks!

**Comment by piyushghai** [comment]
@sxjscience @KellenSunderland Gentle ping...

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Operator]

**Comment by wkcn** [comment]
@sxjscience @KellenSunderland 
Could you please help take a review? Thank you!

File: src/operator/roi_pooling.cc [hunk 1]
```diff
@@ -137,91 +140,18 @@ template<typename Dtype>
...
-              }
-            }
-          }
-          bottom_diff[offset_bottom_diff] += gradient;
-        }
-      }
+  for (int index = 0; index < count; ++index) {
+    index_t max_idx = argmax_data[index];
+    if (max_idx >= 0) {
+      bottom_diff[max_idx] += top_diff[index];
```
Comment (sxjscience) [inline src/operator/roi_pooling.cc]: Is it possible that two threads are writing to the same max_idx?

Comment (sxjscience) [reply src/operator/roi_pooling.cc]: It's correct, sorry for the misreading.

**Comment by sxjscience** [comment]
I think it should be good. One concern is it may be backward incompatible due to the `atomicAdd`. Although I feel it's reasonable to use the faster version, I need to confirm with the GluonCV team whether it will break the current training scripts.

**Comment by zhreshold** [comment]
Would be nice to apply to ROIAlign similarly since GluonCV have transitioned to use ROIAlign in recent networks.

**Comment by larroy** [comment]
This is great, is the performance increase only due to type changes as in the description?

**Comment by wkcn** [comment]
@larroy I think the main reason is that the indices of the maximum numbers are saved in this PR.

**Comment by KellenSunderland** [comment]
LGTM.  Many thanks for the speedup / refactor.
